src/gt4py/next/iterator/tracing.py

```diff
@@ -248,37 +251,82 @@ def closure(domain, stencil, output, inputs):
+    param_infos = list(inspect.signature(fun).parameters.values())
+
+    for i, arg in enumerate(args):
+        if i < len(param_infos):
+            param_info = param_infos[i]
+        else:
+            if param_info.kind != inspect.Parameter.VAR_POSITIONAL:
+                # the last parameter info might also be a keyword or variadic keyword argument, but
+                # they are not supported.
+                raise NotImplementedError(
+                    "Only `POSITIONAL_OR_KEYWORD` or `VAR_POSITIONAL` parameters are supported."
+                )
+            param_info = param_infos[-1]
+
+        if param_info.kind == inspect.Parameter.VAR_POSITIONAL:
+            param_name = f"_{param_info.name}{i}"
+        elif param_info.kind == inspect.Parameter.POSITIONAL_OR_KEYWORD:
+            param_name = param_info.name
+        else:
+            raise NotImplementedError(
+                "Only `POSITIONAL_OR_KEYWORD` or `VAR_POSITIONAL` parameters are supported."
+            )
+
+        kind, dtype = None, None
+        if use_arg_types:
+            # TODO(tehrengruber): Fields of tuples are not supported yet. Just ignore them for now.
+            if not _contains_tuple_dtype_field(arg):
+                arg_type = type_translation.from_value(arg)
+                # TODO(tehrengruber): Support more types.
+                if isinstance(arg_type, type_specifications.FieldType):
+                    kind = "Iterator"
+                    dtype = (
+                        arg_type.dtype.kind.name.lower(),  # actual dtype
+                        type_info.is_local_field(arg_type),  # is list
+                    )
+
+        params.append(Sym(id=param_name, kind=kind, dtype=dtype))
+    return params
+
+
+def trace_fencil_definition(
+    fun: typing.Callable, args: typing.Iterable, *, use_arg_types=True
+) -> FencilDefinition:
+    """
+    Transform fencil given as a callable into `itir.FencilDefinition` using tracing.
+
+    Arguments:
+        fun: The fencil / callable to trace.
+        args: A list of arguments, e.g. fields, scalars or composites thereof. If `use_arg_types`
+            is `False` may also be dummy values.
+
+    Keyword arguments:
+        use_arg_types: Deduce type of the arguments and add them to the fencil parameter nodes
+            (i.e. `itir.Sym`s).
+    """
     with TracerContext() as _:
-        param_names = _make_param_names(fun, args)
-        trace_function_call(fun, args=(_s(p) for p in param_names))
+        params = _make_fencil_params(fun, args, use_arg_types=use_arg_types)
+        trace_function_call(fun, args=(_s(param.id) for param in params))
 
         return FencilDefinition(
             id=fun.__name__,
             function_definitions=TracerContext.fundefs,
-            params=list(Sym(id=param) for param in param_names),
+            params=params,
             closures=TracerContext.closures,
         )
-
-
-def fendef_tracing(fun, *args, **kwargs) -> FencilDefinition:
```

Comment:
This function just dropped the keyword arguments. I removed it and made it explicit in the `FendefDispatcher`.